src/Orleans/IDs/GuidId.cs

```diff
@@ -0,0 +1,124 @@
+/*
+Project Orleans Cloud Service SDK ver. 1.0
+ 
+Copyright (c) Microsoft Corporation
+ 
+All rights reserved.
+ 
+MIT License
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software and 
+associated documentation files (the ""Software""), to deal in the Software without restriction,
+including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
+subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
+OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+﻿using System;
+using Orleans.Concurrency;
+using Orleans.Serialization;
+
+namespace Orleans.Runtime
+{
+    /// <summary>
+    /// Wrapper object around Guid.
+    /// Can be used in places where Guid is optional and in those cases it can be set to null and will not use the storage of an empty Guid struct.
```

Comment:
Guid is a struct. It will take 16 bytes of the grain reference, even if not used.
Are you suggesting Nullable<Guid> instead? I did not think about it.
The advantage of the new type is that it encapsulates all this logic inside it. Like interner. And SerializeToStream and ToParsableString.
If there is any advantage performance-wise of Nullable<Guid> instead of that new class, I will reconsider. Otherwise, I think this is cleaner, since it allows this nice encapsulation.

[as a side note, I am still a bit, deep in my heart, a Java guy. No structs.]